Content.Server/Mining/MineableSystem.cs

```diff
@@ -36,36 +40,57 @@ private void OnInteractUsing(EntityUid uid, MineableComponent component, Interac
         if (!pickaxe.MiningEntities.Add(uid))
             return;
 
-        var doAfter = new DoAfterEventArgs(args.User, component.BaseMineTime * pickaxe.MiningTimeMultiplier, default, uid)
+        component.CancelToken = new CancellationTokenSource();
+        var doAfter = new DoAfterEventArgs(args.User, component.BaseMineTime * pickaxe.MiningTimeMultiplier, component.CancelToken.Token, uid)
         {
             BreakOnDamage = true,
             BreakOnStun = true,
             BreakOnTargetMove = true,
             BreakOnUserMove = true,
             MovementThreshold = 0.5f,
-            BroadcastCancelledEvent = new MiningDoafterCancel() { Pickaxe = args.Used, Rock = uid },
-            BroadcastFinishedEvent = new MiningDoafterSuccess() { Pickaxe = args.Used, Rock = uid }
+            BroadcastCancelledEvent = new MiningDoafterCancel(component) { Pickaxe = args.Used, Rock = uid },
+            TargetFinishedEvent = new MiningDoafterSuccess() { Pickaxe = args.Used, Rock = uid, Player = args.User }
         };
 
         _doAfterSystem.DoAfter(doAfter);
     }
 
-    private void OnDoafterSuccess(MiningDoafterSuccess ev)
+    private void OnDoafterSuccess(EntityUid uid, MineableComponent component, MiningDoafterSuccess ev)
     {
+        var entityManager = IoCManager.Resolve<IEntityManager>();
         if (!TryComp(ev.Pickaxe, out PickaxeComponent? pickaxe))
             return;
 
         _damageableSystem.TryChangeDamage(ev.Rock, pickaxe.Damage);
         SoundSystem.Play(Filter.Pvs(ev.Rock), pickaxe.MiningSound.GetSound(), ev.Rock, AudioParams.Default);
         pickaxe.MiningEntities.Remove(ev.Rock);
+        
+        var spawnOre = EntitySpawnCollection.GetSpawns(component.Ores, _random);
+        foreach (var item in spawnOre)
+        {
+            entityManager.SpawnEntity(item, entityManager.GetComponent<TransformComponent>(ev.Player).Coordinates);
+        }
```

Comment:
This is kinda clunky and should probably try to insert it into their hands. It's also doing multiple transform resolves when it should just do it once.

Reply:
still working on how to spawn in hands, but removed foreach

Reply:
leaving in spawning on ground, as its clunky when mining (you don't always mine the same ore, thus it ends up on the ground anyway)